feat: enhance auth server discovery with OAuth2 and OpenID metadata support

xiaoyijun opened the pull request:
<!-- Provide a brief summary of your changes -->
# Add OpenID Connect Discovery Support for Authorization Server Discovery

Note: this PR reopens https://github.com/modelcontextprotocol/modelcontextprotocol/pull/677

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->

This PR enhances the authorization server discovery mechanism by adding support for OpenID Connect Discovery 1.0 alongside the existing OAuth 2.0 Authorization Server Metadata. 

OpenID Connect (OIDC) is built on top of OAuth 2.0, extending it with standardized identity functionality. Many modern authorization servers implement OIDC as their primary protocol, making it crucial for MCP to support both discovery mechanisms. Popular authorization providers such as Keycloak, Auth0, and Logto all implement OIDC discovery by default.

Key benefits:
- Better compatibility with OIDC-based authorization servers
- Maintains full OAuth 2.0 compatibility since OIDC is an extension of OAuth 2.0
- Enables seamless integration with popular identity providers
- No breaking changes to existing OAuth 2.0 implementations

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->

N/A

## Breaking Changes
<!-- Will users need to update their code or configurations? -->

None. This change is fully backwards compatible:
- Existing OAuth 2.0 metadata discovery continues to work as before
- Authorization servers can choose which discovery mechanism to implement
- MCP clients must support both mechanisms but will automatically use the correct one

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

The addition of OIDC discovery support is particularly valuable because:
1. OIDC is the de-facto standard for modern identity providers
2. The `.well-known/openid-configuration` endpoint is widely supported
3. OIDC providers automatically support OAuth 2.0 flows since OIDC is built on top of OAuth 2.0
4. This enables MCP to work seamlessly with popular authorization servers like Keycloak, Auth0, and Logto without additional configuration

joshcanhelp commented:
$0.02 on this ... I 100% support the OIDC metadata fallback but [RFC8414 already mentions this fallback](https://datatracker.ietf.org/doc/html/rfc8414#section-3). Supporting that RFC fully would mean that the fallback support is required and wouldn't need to be spelled out explicitly. I think the spec could say something to the effect of:

> **MUST** support RFC8414 to provide at least one mechanism for obtaining AS metadata 

... or thereabouts.

Related PRs/discussions:

- https://github.com/modelcontextprotocol/typescript-sdk/pull/585
- https://github.com/modelcontextprotocol/modelcontextprotocol/issues/506

xiaoyijun commented:
Hi @joshcanhelp thank you for the feedback and suggestions.

However, I believe that treating “support for RFC 8414” as a blanket requirement misses a key architectural consideration, for the following reasons:

As you noted, Section 3 of RFC 8414 states:
> Different applications utilizing OAuth authorization servers in application-specific ways may define and register different well-known URI suffixes used to publish authorization server metadata as used by those applications.

The spec mentions that for an auth server used by a specific application (e.g., "example"), its metadata endpoint could be `/.well-known/example-configuration`. This is for **application-specific** scenarios. This approach introduces a new challenge: you would need a clearer way for developers to know how to construct this URL from an application's identifier, which would require *more* specification work to guide auth server discovery.

For the MCP spec, considering its goal of ecosystem compatibility (to adapt to general-purpose auth servers), the more appropriate choice is the default discovery endpoint, which RFC 8414 also mentions:
> ...many such applications will use the default well-known URI string “/.well-known/oauth-authorization-server”, which is the right choice for general-purpose OAuth authorization servers, and not register an application-specific one.

Regarding `/.well-known/openid-configuration`, the RFC states:
> Some OAuth applications will choose to use the well-known URI suffix “openid-configuration”. As described in Section 5, despite the identifier “/.well-known/openid-configuration”, appearing to be OpenID specific, its usage in this specification is actually referring to a general OAuth 2.0 feature that is not specific to OpenID Connect.

So, from the perspective of RFC 8414, there isn't really an "OIDC concept" at play here. The name `openid-configuration` is just a string that happens to fit the `example-configuration` pattern.

This distinction frames our decision as a choice between two architectural paths:

1.  **(The path I rejected):** The `application-specific` route. This would introduce unnecessary complexity by forcing us to define how clients discover these custom URLs.
2.  **(The path I chose):** Supporting a widely adopted, de facto industry standard (OIDC Discovery). This is a pragmatic and robust choice for our ecosystem.

This PR deliberately chooses the second path.

@localden @aaronpk @D-McAdams I'd be interested to hear your point of view on this.

joshcanhelp commented:
At the end of the day, we want these applications to be able to get the authorization server metadata so they're able to log in. Whether it's one of specific URLs stated or an application-specific one doesn't matter, in my mind. I don't think we should be deciding that one of those two paths, both of which are currently being used, is the right way when the goal is the same. The question, in my mind, is ... does the application-specific path limit this in any way or affect developers that aren't using it? The RFC solves the problem, I suggest we just point there and leave it at that. 

Appreciate the thoughtful conversation, either way!

pwwpche commented:
> At the end of the day, we want these applications to be able to get the authorization server metadata so they're able to log in. Whether it's one of specific URLs stated or an application-specific one doesn't matter. I don't think we should be deciding that one of those two paths, both of which are currently being used, is the right way when the goal is the same. The question, in my mind, is ... does allowing the application-specific path limit this in any way or affect developers that aren't using it? The RFC solves the problem, I suggest we just point there and leave it at that.
> 
> Appreciate the thoughtful conversation, either way!

If I understand correctly from RFC 8414:
> The well-known URI suffix used MUST be registered in the IANA "Well-Known URIs" registry

This limits that even for application-specific URLs, they have to be listed in [IANA](https://www.iana.org/assignments/well-known-uris/well-known-uris.xml). Also regarding allowing application specific path (that may not be on the IANA list), could you suggest a few use cases? I'm struggling to understand the value it provides, given that `OIDC is the de-facto standard for modern identity providers`.

pwwpche commented:
@xiaoyijun 
This looks great to me, thanks for putting it up!

Regarding a remaining question from [previous PR](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/677):

> how clients should handle the difference in fields between OICD and OAuth 2.0 AS metadata

AS Metadata and OIDC config requires different fields. See the comparison that @localden shared:
https://gist.github.com/localden/26d8bcf641703c08a5d8741aa9c3336c

I've took a glance and it doesn't seem to me that supporting OIDC requires handling the difference. However could you help with a second eye on checking if any of the differences are required by Auth implementations like https://github.com/modelcontextprotocol/python-sdk/commit/17f9c00c53b4463ac1a449d601c2d181664ab857?

LucaButBoring commented on docs/specification/draft/basic/authorization.mdx:
```diff
@@ -63,9 +63,12 @@ specifies how an MCP server indicates the location of its corresponding authoriz
+   - OAuth 2.0 Authorization Server Metadata ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
+   - [OpenID Connect Discovery 1.0](https://openid.net/specs/openid-connect-discovery-1_0.html)
+
+   MCP clients **MUST** support both discovery mechanisms to obtain the information required to interact with the authorization server.
```
Do we need guidance on how to support both simultaneously in a client?

For example, is it safe for an SDK to automatically fall back from RFC8414 to OIDC 1.0 if one isn't present? Is the discovery mechanism allowed to be implicit at all, or should it always be explicit? Does it matter which one we check first, and is it acceptable for SDKs to diverge in that behavior?

SDKs will likely diverge here if this isn't specified, but I'm not sure if that's acceptable or not.

xiaoyijun replied:
Hi @LucaButBoring , thanks for your thorough analysis.

You're absolutely right that the current specification creates ambiguity around client implementation, which could lead to interoperability issues between different SDKs.

I'm currently working on updating the specification to cover the metadata discovery flow more comprehensively.

softnado commented:
$0.02 on the subject.
Current auth server discovery sequence require /.well-known/oauth-authorization-server be hosted under Auth Server domain.
Which is not supported and isn't practical in my opinion for enterprise auth service like Microsoft Entra.
For such centralized and shared auth service, it is more practical for resource owner to host the /.well-known/oauth-authorization-server or alternative metadata and ensure its correctness for discovery.

Puls, it opens the possibilities for Resource Server to provide its own Dynamic Client Registration Endpoint while other Endpoints remain pointed to the enterprise auth service.
It reduces the need of create wrapping or re-create Auth Endpoints under Resource Server domain (Which usually isn't allowed under enterprise scenario). And the custom Dynamic Client Registration Endpoint give more control to Resource Owner for auditing, also unblock scenario the auth service doesn't support Dynamic Client Registration yet. (In that situation, the client returned can be proactively created and permission scoped)

Looking forward to learning more about your thoughts and other common enterprise auth service support of this scenario.

softnado commented:
> > For such centralized and shared Auth Server, it is more practical for Resource Server owner to host the /.well-known/oauth-authorization-server or alternative metadata and ensure its correctness for discovery.
> 
> Please Correct me if Im missing something, but isn't this already covered by the protected resource metadata spec[1]?
> 
> [1] https://datatracker.ietf.org/doc/rfc9728/

It depends on the Auth Server required by Enterprise Context, for Auth Server like Microsoft Entra, there is not support for /.well-known/oauth-authorization-server, and personally I believe that is not practical for such Auth Server which targeted for multi tenants/enterprises.

The current spec's assumption of Auth Server I believe assumed the Auth Server have strong flexibility to the Resource Server owner which isn't the case for Enterprise Scenarios.

LucaButBoring commented:
> Current Auth Server discovery sequence require /.well-known/oauth-authorization-server be hosted under Auth Server domain.

@softnado can you give a specific example of the problem, including what you'd like the `issuer` to look like? IIUC this isn't a problem today - if you want to use metadata from another server, simply include that server as the `issuer`.

Also, unless I'm missing something, that's not OIDC-related, so we can spin that off to another RFC if there is a gap there.

LucaButBoring commented on docs/specification/draft/basic/authorization.mdx:
```diff
@@ -93,8 +96,20 @@ MCP clients **MUST** be able to parse `WWW-Authenticate` headers and respond app
+
+The discovery approach is based on [RFC8414 Section 3.1 "Authorization Server Metadata Request"](https://datatracker.ietf.org/doc/html/rfc8414#section-3.1) for OAuth 2.0 Authorization Server Metadata discovery and [RFC8414 Section 5 "Compatibility Notes"](https://datatracker.ietf.org/doc/html/rfc8414#section-5) for OpenID Connect Discovery interoperability.
+
+For issuer URLs with path components (e.g., `https://auth.example.com/tenant1`), clients **MUST** try endpoints in the following priority order:
+
+1. OAuth 2.0 Authorization Server Metadata with path insertion: `https://auth.example.com/.well-known/oauth-authorization-server/tenant1`
+2. OpenID Connect Discovery with path insertion: `https://auth.example.com/.well-known/openid-configuration/tenant1`
+3. OpenID Connect Discovery 1.0 legacy path appending: `https://auth.example.com/tenant1/.well-known/openid-configuration`
+
+For issuer URLs without path components (e.g., `https://auth.example.com`), clients **MUST** try:
+
+1. OAuth 2.0 Authorization Server Metadata: `https://auth.example.com/.well-known/oauth-authorization-server`
+2. OpenID Connect Discovery: `https://auth.example.com/.well-known/openid-configuration`
```
Looks good, just one minor nit:
> OpenID Connect Discovery 1.0 legacy path appending

I'd remove the `legacy` from this phrasing, actually - RFC8414 describes that as legacy in the context of OAuth 2.0, but given that the prior phrasing in this PR is...
> MCP authorization servers **MUST** provide at least one of the following discovery mechanisms:
> 
> - OAuth 2.0 Authorization Server Metadata ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
> - [OpenID Connect Discovery 1.0](https://openid.net/specs/openid-connect-discovery-1_0.html)

...that means OIDC is also being treated as a first-class citizen in its own right, now. So, given that OAuth 2.0 does not supercede OIDC, the OIDC semantics are now perfectly valid on their own, too.

xiaoyijun replied:
Good call, I have removed the `legacy` expression.

LucaButBoring commented:
One general question - does supporting OIDC for metadata discovery also imply that clients should support the full range of OIDC authentication semantics? In particular, I'm concerned about `code_challenge_methods_supported` in RFC 8414, which is required when the AS supports PKCE in `authorization_code` flows (which it [must](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#authorization-code-protection)). OIDC does not define the `code_challenge_methods_supported` field (or indeed PKCE at all), meaning that OIDC authorization servers cannot be used in MCP unless they implicitly follow RFC 8414 already.

---

Linking @localden's [schema comparison](https://gist.github.com/localden/26d8bcf641703c08a5d8741aa9c3336c) for reference.

xiaoyijun commented:
I just rebased this PR to fix lint issues in other files that are unrelated to this PR.
I also removed the 'legacy' description from the OIDC path appending method based on @LucaButBoring  suggestion.

xiaoyijun commented:
> One general question - does supporting OIDC for metadata discovery also imply that clients should support the full range of OIDC authentication semantics? In particular, I'm concerned about `code_challenge_methods_supported` in RFC 8414, which is required when the AS supports PKCE in `authorization_code` flows (which it [must](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#authorization-code-protection)). OIDC does not define the `code_challenge_methods_supported` field (or indeed PKCE at all), meaning that OIDC authorization servers cannot be used in MCP unless they implicitly follow RFC 8414 already.
> 
> Linking @localden's [schema comparison](https://gist.github.com/localden/26d8bcf641703c08a5d8741aa9c3336c) for reference.

Hi @LucaButBoring, thanks for raising this important point! You're absolutely right to be concerned about `code_challenge_methods_supported` and PKCE compatibility.

I did some research into this issue, and here's what I found:

## The Core Problem:
- OAuth 2.1 and PKCE specs require authorization servers to support PKCE, but they don't actually define any mechanism for clients to verify this support
- The only reliable way to discover PKCE support is `through the code_challenge_methods_supported` field in Authorization Server Metadata ([RFC 8414 section 2](https://datatracker.ietf.org/doc/html/rfc8414#section-2))
>  `code_challenge_methods_supported`: JSON array containing a list of Proof Key for Code
      Exchange (PKCE) [[RFC7636](https://datatracker.ietf.org/doc/html/rfc7636)] code challenge methods supported by this
      authorization server. ... If omitted, the authorization server does not support PKCE.
- OIDC Discovery spec doesn't define this field, creating a gap for OIDC-only providers

## Current Reality:
- Most OIDC providers already include code_challenge_methods_supported in their discovery responses as an extension ([Google](https://accounts.google.com/.well-known/openid-configuration), [GitLab](https://gitlab.com/.well-known/openid-configuration), [HuggingFace](https://huggingface.co/.well-known/openid-configuration), [Auth0](https://dev-utwzif1y.us.auth0.com/.well-known/openid-configuration), [Logto](https://6gursn.app.logto.dev/oidc/.well-known/openid-configuration) etc.)
- Since OIDC is an identity layer built on OAuth 2.0, it should logically follow the same discovery patterns
- There's no other reliable mechanism to verify PKCE support (probing with requests is unreliable since servers may silently ignore unknown parameters)

## My proposed solution:
Since OIDC is based on OAuth 2.0, there's no alternative reliable discovery mechanism, and MCP absolutely requires PKCE for security, I'm planning to update the spec to require:
- OIDC providers offering MCP compatibility MUST include `code_challenge_methods_supported` in their discovery responses
- MCP clients MUST verify this field is present.
- This aligns OIDC discovery with OAuth 2.0 Authorization Server Metadata practices

This ensures we never compromise on PKCE security while providing a clear, implementable standard for OIDC providers.

What do you think of this approach?

@LucaButBoring, cc @localden @D-McAdams @dsp-ant

LucaButBoring commented:
> Since OIDC is based on OAuth 2.0, there's no alternative reliable discovery mechanism, and MCP absolutely requires PKCE for security, I'm planning to update the spec to require:
> 
> OIDC providers offering MCP compatibility MUST include code_challenge_methods_supported in their discovery responses
> - MCP clients MUST verify this field is present.
> - This aligns OIDC discovery with OAuth 2.0 Authorization Server Metadata practices
> - This ensures we never compromise on PKCE security while providing a clear, implementable standard for OIDC providers.

Yes, this sounds reasonable to me - aligned with this way of handling it for the time being.

000-000-000-000-000 commented:
> @xiaoyijun Do we have implementations for these in Python or Typescript? Per the new process: https://modelcontextprotocol.io/community/sep-guidelines we really would love to see at least one implementation

@dsp-ant https://github.com/modelcontextprotocol/python-sdk/pull/1061

TylerLeonhardt commented on docs/specification/draft/basic/authorization.mdx:
```diff
@@ -93,8 +96,20 @@ MCP clients **MUST** be able to parse `WWW-Authenticate` headers and respond app
+
+For issuer URLs with path components (e.g., `https://auth.example.com/tenant1`), clients **MUST** try endpoints in the following priority order:
+
+1. OAuth 2.0 Authorization Server Metadata with path insertion: `https://auth.example.com/.well-known/oauth-authorization-server/tenant1`
+2. OpenID Connect Discovery with path insertion: `https://auth.example.com/.well-known/openid-configuration/tenant1`
+3. OpenID Connect Discovery 1.0 path appending: `https://auth.example.com/tenant1/.well-known/openid-configuration`
```
isn't path appending more "spec aligned" for OIDC? Shouldn't we do the "correct" method first, and _then_ fallback to the path insertion?

In other words, I'm wondering if 2 and 3 should be swapped.

dsp-ant commented:
This was accepted by @modelcontextprotocol/core-maintainers with a 7 YES votes, 0 NO votes and 1 ABSENT

xiaoyijun commented:
> @xiaoyijun Do we have implementations for these in Python or Typescript? Per the new process: https://modelcontextprotocol.io/community/sep-guidelines we really would love to see at least one implementation

Hi @dsp-ant , yes, I have a related WIP TypeScript SDK PR here: https://github.com/modelcontextprotocol/typescript-sdk/pull/652

Now let this spec is basically settled now, I'll update the SDK implementation to match these new requirements. Will let you know when it's ready!
